### PR TITLE
Feature/tmi2 182 create api key

### PR DESCRIPTION
**Pull request by a-lor-cab**

### CODE REVIEW

File: src/main/resources/application.properties [hunk 1]
```diff
@@ -3,7 +3,7 @@ spring.datasource.username=postgres
 spring.datasource.password=mysecretpassword
 spring.datasource.hikari.maximumPoolSize=5
 
-server.port=8080
+server.port=8084
```
Comment (ConnorTCO) [inline src/main/resources/application.properties:6]: will this port change affect anything in aws?

File: src/main/java/gov/cabinetoffice/gapfindapiadmin/controllers/ApiKeyController.java [hunk 2]
```diff
@@ -1,24 +1,65 @@
...
+                    null,
+                    "An API key with this name already exists");
+            bindingResult.addError(duplicateKey);
+        }
+        if (bindingResult.hasErrors()) {
+            return new ModelAndView(CREATE_API_KEY_FORM_PAGE)
+                    .addObject("createApiKeyDTO", createApiKeyDTO);
+        }
+
+        final ModelAndView newApiKey = new ModelAndView(NEW_API_KEY_PAGE);
```
Comment (GavCookCO) [inline src/main/java/gov/cabinetoffice/gapfindapiadmin/controllers/ApiKeyController.java:61]: I don't think this needs to be 3 lines of code but thats just a personal style preference

File: src/main/java/gov/cabinetoffice/gapfindapiadmin/controllers/controller_advice/ControllerExceptionsHandler.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+import software.amazon.awssdk.services.apigateway.model.ApiGatewayException;
+
+@ControllerAdvice(
+        assignableTypes = {ApiKeyController.class})
+public class ControllerExceptionsHandler extends ResponseEntityExceptionHandler {
+
+
+    @ExceptionHandler(value = {ApiGatewayException.class})
+    protected ResponseEntity<Object> handleException(ApiGatewayException ex, WebRequest request) {
```
[the remark this answers is not in the document]
Comment (GavCookCO) [reply src/main/java/gov/cabinetoffice/gapfindapiadmin/controllers/controller_advice/ControllerExceptionsHandler.java:21]: Would suggest that ControllerAdvice classes are more suitable for REST APIs but please feel free to do some more reading into how they can be used with Thymeleaf and Spring MVC

File: src/main/java/gov/cabinetoffice/gapfindapiadmin/repositories/GrantAdminRepository.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package gov.cabinetoffice.gapfindapiadmin.repositories;
+
+import gov.cabinetoffice.gapfindapiadmin.models.GrantAdmin;
+import org.springframework.data.repository.CrudRepository;
+
+public interface GrantAdminRepository extends CrudRepository<GrantAdmin, Integer> {
+    GrantAdmin findByGapUserUserSub(String userSub);
```
Comment (GavCookCO) [inline src/main/java/gov/cabinetoffice/gapfindapiadmin/repositories/GrantAdminRepository.java:7]: Some of the repositories are returning optional objects and some are returning concrete objects for the same kinds of method calls. Is there a reason for that? Could we decide on a consistent approach and stick to it?

File: src/main/resources/application.properties (hunk 1 above)
Comment (a-lor-cab) [reply src/main/resources/application.properties:6]: no, but when u test locally, with the admin running on 8080, and user service on 8082, 8084 was the natural call for it 😅

File: src/main/java/gov/cabinetoffice/gapfindapiadmin/controllers/controller_advice/ControllerExceptionsHandler.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (a-lor-cab) [reply src/main/java/gov/cabinetoffice/gapfindapiadmin/controllers/controller_advice/ControllerExceptionsHandler.java:21]: implemented an error-page for it

File: src/main/java/gov/cabinetoffice/gapfindapiadmin/repositories/GrantAdminRepository.java (hunk 4 above)
Comment (a-lor-cab) [reply src/main/java/gov/cabinetoffice/gapfindapiadmin/repositories/GrantAdminRepository.java:7]: removed the only optional we had, and added a null check

File: src/main/resources/templates/new-api-key.html [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+<body class="govuk-template__body">
+<div th:replace="~{fragments/header :: header}"></div>
+
+<div class="govuk-width-container">
+    <main class="govuk-main-wrapper govuk-main-wrapper--l" id="main-content" role="main">
+        <div class="govuk-grid-row">
+            <div class="govuk-grid-column-two-thirds">
+                <h1 class="govuk-heading-l">New API key</h1>
+                <p class="govuk-heading-l">Your API key:</p>
+                <p class="govuk-heading-l" th:text="${keyValue}" id="apiKeyValue"></p>
```
Comment (rachelswart) [inline src/main/resources/templates/new-api-key.html:19]: These two should have class="govuk-heading-m"

**Review by rachelswart — APPROVED** [review]

**Pull request merged**